CARDS-2256: PREMs Overwrite LENGTH_OF_STAY_DAYS with calculation

acrowthe opened the pull request:
- Added configuration to LengthOfStayMapper: Overwrite existing length by default
- Changed length of stay logic to ignore time of day when calculating stay duration
- Re-added Discard-IP-ShortStay discard filter

If you encounter issues running generate_compose_yaml due to a tzlocal error, there is a [test branch here](https://github.com/data-team-uhn/cards/tree/misc-prems-length-of-stay-test) which may fix your issue.

Testing instructions
1. Build with `-Pdocker` and navigate to `compose-cluster`
2. Clean up any old CARDS docker images: `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
3. Prepare docker with the following flags: `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --composum --adminer`
4. Start docker: `docker-compose build && docker-compose up -d`
5. Nacigate to http://localhost:1435/ and sign in:
- System: `MS SQL (beta)`
- Server: `mssql`
- Username: `sa`
- Password: `testPassword_`
- Database: `master`
6. Select Import on the left bar
7. Upload `cards/compose-cluster/mssql/processorImportTest.sql` and execute. It should say `38 queries executed OK`
8. Login to CARDS as admin
9. Navigate to http://localhost:8080/Subjects.importClarity. A mostly empty page should show success and Data import started
10. Navigate to the daashboard http://localhost:8080/content.html/Questionnaires/User and check which of patients 34 through 36 were imported.

Expected result:
Patients 34 and 35 should not be imported due to a length of stay of less than 1 day (10 hours or 23 hours 58 minutes)
Patient 36 should be imported as an inpatient or long form inpatient due to an 11 minute visit crossing midnight

11. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
12. Repeat steps 3 through 8
13. Delete the configuration for LengthOfStayMapper using `http://localhost:8080/system/console/configMgr`
14. Repeat steps 9 and 10. Verify the results are the same as before due to default handling if configuration is not present.
15. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
16. Repeat steps 3 through 8
17. Modify the configuration for LengthOfStayMapper using `http://localhost:8080/system/console/configMgr` to disable overwrite
14. Repeat steps 9 and 10. Verify that all of patients 34-36 are present as imported default LENGTH_OF_STAY of 1 is accepted

acrowthe committed:
CARDS-2256: PREMs LENGTH_OF_STAY_DAYS Improvements
- Added configuration to LengthOfStayMapper: Overwrite existing length by default
- Changed length of stay logic to ignore time of day when calculating stay duration
- Re-added Discard-IP-ShortStay discard filter

marta- commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -363,6 +363,11 @@
+    // Discard patients that were only in-patient for a short time
+    "io.uhndata.cards.clarity.importer.internal.ConfiguredDiscardFilter~Discard-IP-ShortStay":{
+      "priority": 100,
+      "conditions": ["CLINIC is empty", "LENGTH_OF_STAY_DAYS <= 4"]
```
```suggestion
      "conditions": ["CLINIC is empty", "LENGTH_OF_STAY_DAYS < 1"]
```

acrowthe committed:
CARDS-2256: PREMs LENGTH_OF_STAY_DAYS Improvements

Adjust filter down to 1 day from 4

Co-authored-by: Marta Girdea <marta@phenotips.org>

acrowthe committed:
CARDS-2256: PREMs LENGTH_OF_STAY_DAYS Improvements
- Update test cases

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -363,6 +363,11 @@
       "conditions": ["CLINIC is empty", "DISCH_DEPT_NAME matches .*EMERGENCY.*"]
     },
     // Only in-patients from now on.
```
@marta- I'm not sure this will actually happen with real data, but what should in theory happen to rehab patients with a length of stay shorter than 1? As things are in this PR, only IP and EDIP with short stay get discarded, meaning that a rehab patient with a few hours stay still receives the Rehab survey.

acrowthe replied:
Added new `Discard-Rehab-ShortStay` discard filter to discard rehab patients with <1 day visits

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/LengthOfStayMapper.java:
```diff
@@ -55,14 +78,17 @@ public Map<String, String> processEntry(Map<String, String> input)
                 admission.setTime(dateFormat.parse(input.getOrDefault("HOSP_ADMISSION_DTTM", "")));
                 discharge.setTime(dateFormat.parse(input.getOrDefault("HOSP_DISCHARGE_DTTM", "")));
 
+                admission = DateUtils.truncate(admission, Calendar.DATE);
+                discharge = DateUtils.truncate(discharge, Calendar.DATE);
```
I think this should be configurable as well. Add a new boolean setting in the configuration, `useCalendarDays`, default to true, and wrap these two lines in an if on that setting.

sdumitriu commented:
Oh, and I think this deserves a Jira issue, please create one and amend the commit and PR title.

acrowthe committed:
CARDS-2256: PREMs LENGTH_OF_STAY_DAYS Improvements
- Discard short stay rehab
- Add in calendar days vs. 24 hour day configuration

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -311,6 +311,11 @@
       "clinic": "/Survey/ClinicMapping/78840662",
       "conditions": ["DISCH_LOC_NAME matches Toronto Rehab .*"]
     },
+    // Discard rehab patients that were only for a short time
```
```suggestion
    // Discard rehab patients that were only in-patient for a short time
```

sashaandjic reviewed (approved):
tested

sdumitriu committed:
[misc] Better wording

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -363,6 +368,11 @@
+    // Discard patients that were only in-patient for a short time
+    "io.uhndata.cards.clarity.importer.internal.ConfiguredDiscardFilter~Discard-IP-ShortStay":{
+      "priority": 100,
+      "conditions": ["CLINIC is not /Survey/ClinicMapping/-1792626799", "LENGTH_OF_STAY_DAYS < 1"]
```
How did nobody notice that `is not` is not a known operator?

sdumitriu replied:
And if it is corrected to `CLINIC <> /Survey/ClinicMapping/-1792626799`, then it applies to all clinics other than ED, so it makes the rehab-specific one redundant.

marta- replied:
Making a record here of the fact that this issue is being addressed in  #1530
